File: Python/0108-e-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        if not nums:
            return None
        head = self.helper(nums, 0, len(nums) - 1)
        return head
    
    def helper(self,nums, low, high):
        if low > high:
            return None
        mid = (low + high) / 2
        node = TreeNode(nums[mid])
        node.left = self.helper(nums, low, mid - 1)
        node.right = self.helper(nums, mid + 1, high)
        return node

File: leetcode/C++/0108-e-convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return helper(nums.begin(), nums.end());
    }
private:
    TreeNode* helper(vector<int>::const_iterator first, vector<int>::const_iterator last) {
        const auto length = distance(first, last);

        if (length <= 0) return nullptr;

        auto mid = first + length / 2;
        TreeNode* root = new TreeNode(*mid);
        root->left = helper(first, mid);
        root->right = helper(mid + 1, last);

        return root;
    }
};